lib/Drush/Boot/DrupalBoot.php

```diff
@@ -2,115 +2,471 @@
+      DRUSH_BOOTSTRAP_DRUPAL_ROOT            => 'bootstrap_drupal_root',
+      DRUSH_BOOTSTRAP_DRUPAL_SITE            => 'bootstrap_drupal_site',
+      DRUSH_BOOTSTRAP_DRUPAL_CONFIGURATION   => 'bootstrap_drupal_configuration',
+      DRUSH_BOOTSTRAP_DRUPAL_DATABASE        => 'bootstrap_drupal_database',
+      DRUSH_BOOTSTRAP_DRUPAL_FULL            => 'bootstrap_drupal_full',
+      DRUSH_BOOTSTRAP_DRUPAL_LOGIN           => 'bootstrap_drupal_login');
+  }
+
+  function bootstrap_init_phases() {
+    return array(DRUSH_BOOTSTRAP_DRUSH, DRUSH_BOOTSTRAP_DRUPAL_FULL);
+  }
+
+  function enforce_requirement(&$command) {
+    parent::enforce_requirement($command);
+    $this->drush_enforce_requirement_drupal_dependencies($command);
+  }
+
+  function report_command_error($command) {
+    // If we reach this point, command doesn't fit requirements or we have not
+    // found either a valid or matching command.
 
-    drush_set_context('DRUSH_BOOTSTRAP_PHASE', DRUSH_BOOTSTRAP_NONE);
+    // If no command was found check if it belongs to a disabled module.
+    if (!$command) {
+      $command = $this->drush_command_belongs_to_disabled_module();
+    }
+    parent::report_command_error($command);
   }
 
   function command_defaults() {
     return array(
+      'drupal dependencies' => array(),
       'bootstrap' => DRUSH_BOOTSTRAP_DRUPAL_LOGIN,
     );
   }
 
-  function bootstrap_and_dispatch() {
-    $phases = _drush_bootstrap_phases(FALSE, TRUE);
-
-    $return = '';
-    $command_found = FALSE;
-    _drush_bootstrap_output_prepare();
-    foreach ($phases as $phase) {
-      if (drush_bootstrap_to_phase($phase)) {
-        $command = drush_parse_command();
-        if (is_array($command)) {
-          $command += $this->command_defaults();
-          $bootstrap_result = drush_bootstrap_to_phase($command['bootstrap']);
-          drush_enforce_requirement_bootstrap_phase($command);
-          drush_enforce_requirement_core($command);
-          drush_enforce_requirement_drupal_dependencies($command);
-          drush_enforce_requirement_drush_dependencies($command);
-
-          if ($bootstrap_result && empty($command['bootstrap_errors'])) {
-            drush_log(dt("Found command: !command (commandfile=!commandfile)", array('!command' => $command['command'], '!commandfile' => $command['commandfile'])), 'bootstrap');
-
-            $command_found = TRUE;
-            // Dispatch the command(s).
-            $return = drush_dispatch($command);
-
-            // Prevent a '1' at the end of the output.
-            if ($return === TRUE) {
-              $return = '';
-            }
-
-            if (drush_get_context('DRUSH_DEBUG') && !drush_get_context('DRUSH_QUIET')) {
-              // @todo Create version independant wrapper around Drupal timers. Use it.
-              drush_print_timers();
-            }
-            break;
+  function commandfile_searchpaths($phase, $phase_max = FALSE) {
```

Comment:
This still has a lot of variation by major version. Would be worth splitting this up in the future, IMO.